### ISSUE TITLE
Binding a pipeline with no dynamic state followed by a pipeline with dynamic states

### ISSUE THREAD

**Issue by ntsh-oni** [KhronosGroup/Vulkan-Docs#2276]
The specification says:

> Instead, dynamic state can be modified at any time and persists for the lifetime of the command buffer, or until modified by another dynamic state setting command, or made invalid by another pipeline bind with that state specified as static.

[10.11. Dynamic State](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#pipelines-dynamic-state)

From my understanding, the first quote means that if you use a dynamic state setting command (let's consider [vkCmdSetViewport](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkCmdSetViewport.html) and *VK_DYNAMIC_STATE_VIEWPORT* for this issue, but it should be applicable to all dynamic states), it:
- is valid during the command buffer lifetime,
- is invalidated by calling vkCmdSetViewport again,
- is invalidated by binding a pipeline with a static viewport state (*VK_DYNAMIC_STATE_VIEWPORT* not used).

By doing:
- Binding a pipeline with a static viewport state (*VK_DYNAMIC_STATE_VIEWPORT* not used), then calling vkCmdSetViewport (and drawing), is an error covered by valid usages in [VUID-vkCmdDraw-None-08608](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkCmdDraw-None-08608) (and equivalent for vkCmdDrawIndexed, etc.),
- Binding a pipeline with a dynamic viewport state (*VK_DYNAMIC_STATE_VIEWPORT* used) but never calling vkCmdSetViewport (and drawing), is an error covered by valid usages in [VUID-vkCmdDraw-None-07831](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkCmdDraw-None-07831) (and equivalent for vkCmdDrawIndexed, etc.),
- Calling vkCmdSetViewport, then binding a pipeline with a static viewport state (*VK_DYNAMIC_STATE_VIEWPORT* not used) (and drawing), then binding a pipeline with a dynamic viewport state (*VK_DYNAMIC_STATE_VIEWPORT* used) (and drawing), is an error, as the specification states:

> If the state is specified as dynamic in the new pipeline object, then that command buffer state is not disturbed. Before any draw or dispatch call with this pipeline there must have been at least one call to each of the corresponding dynamic state setting commands. The state-setting commands must be recorded after command buffer recording was begun, or **after the last command binding a pipeline object with that state specified as static**, whichever was the latter.

but does not seem to be covered by valid usage, and is then not caught by validation layers.

[VUID-vkCmdDraw-None-07831](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkCmdDraw-None-07831) is not wrong but seems incomplete, as calling the pipeline state setting command in the command buffer before drawing is not the only condition to be a valid draw call, as it must also be called after the last command binding a pipeline with this state as static.

To sum this up, using a pipeline state setting command, then binding a pipeline with this state as static, then binding a pipeline with this state as dynamic is against the specification but is missing from valid usages and not caught by validation layers.

**Comment by pdaniell-nv** [comment]
In summary, I think the issue you're reporting is that there are no VU statements enforcing the specification text
> The state-setting commands must be recorded after command buffer recording was begun, or after the last command binding a pipeline object with that state specified as static, whichever was the latter.

For example:
<dl>
<dt><a href="https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkCmdDraw-None-07831">VUID-vkCmdDraw-None-07831</a></dt>
<dd>If the bound graphics pipeline state was created with the VK_DYNAMIC_STATE_VIEWPORT dynamic state enabled then
<a href="https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#vkCmdSetViewport">vkCmdSetViewport</a> must have been called in the current command buffer prior to this drawing command</dd>
</dl>
misses the part where vkCmdSetViewport needs to be called after the last graphics pipeline bind with a pipeline object that does not have VK_DYNAMIC_STATE_VIEWPORT set.

I think you're correct and this needs to be fixed. It's quite possible that VVL implemented this correctly anyway and the issue is limited to the specification VU statements.

**Comment by ZaOniRinku** [comment]
I didn't try with the latest SDK yet but when I started experimenting with it (November 10th), cases 1 and 2 triggered validation but case 3 did not.

**Comment by pdaniell-nv** [comment]
Okay, once the VUs are fixed we'll follow up with VVL fixes to match.

**Comment by spencer-lunarg** [comment]
I think this really was a validation layer issue, the change has been merged and now will correctly catch this (and also provide a useful error message like

> vkCmdDraw(): VK_DYNAMIC_STATE_PRIMITIVE_RESTART_ENABLE state is dynamic, but the command buffer never called vkCmdSetPrimitiveRestartEnable. (There was a call to vkCmdBindPipeline with VkPipeline 0x27d60e0000000019[] that didn't have the dynamic state and invalidated the prior vkCmdSetPrimitiveRestartEnable call).

We are update the VUs internally to be more explicit/clear, but the spec itself is already correct (as pointed out in this issue above) so I feel this issue can be closed

**Comment by oddhack** [comment]
The VU updates should be in 1.3.286 spec update. Closing.
